### ISSUE TITLE
take_until! returns an error while take_until_either! works

### ISSUE THREAD

**Issue by seppo0010** [Geal/nom#433]
I'm writing a parser for an xml-like input.

Here's my code:

```rust
#[macro_use]
extern crate nom;

use nom::alphanumeric;

named!(open_tag, ws!(delimited!(tag!("<"), alphanumeric, tag!(">"))));
named!(close_tag, ws!(delimited!(tag!("</"), alphanumeric, tag!(">"))));
named!(self_closing_tag, ws!(delimited!(tag!("<"), alphanumeric, tag!("/>"))));

named!(nodes<&[u8], Vec<Node> >, many0!(node));

named!(tag<Node>, ws!(do_parse!(t: open_tag >> c: nodes >> close_tag >> (Node::Tag(String::from_utf8(t.to_vec()).unwrap(), vec![], c)))));
named!(content<&[u8], &[u8] >, take_until!("<")); // this fails
// named!(content<&[u8], &[u8] >, take_until_either!("<")); // this works fine

named!(node<Node>, alt!(
      self_closing_tag => { |x: &[u8]| Node::Tag(String::from_utf8(x.to_vec()).unwrap(), vec![], vec![]) }
    | tag => { |x| x }
    | content => { |x: &[u8]| Node::Text(String::from_utf8(x.to_vec()).unwrap()) }
));

#[test]
fn test_text() {
    assert_eq!(nodes(b"<a>hello</a>").unwrap(), (&b""[..], vec![Node::Tag("a".to_owned(), vec![], vec![Node::Text("hello".to_owned())])]));
}

#[derive(Debug, Clone, PartialEq)]
enum Node {
    Tag(String, Vec<(String, String)>, Vec<Node>),
    Text(String),
}
```

I'm getting the following output when running the test:

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running target/debug/deps/nom_test-3862c50a7daade96

running 1 test
test test_text ... FAILED

failures:

---- test_text stdout ----
        thread 'test_text' panicked at 'unwrap() called on an IResult that is Error', /home/seppo/.cargo/registry/src/github.com-1ecc6299db9ec823/nom-2.0.1/src/internal.rs:154
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    test_text

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured

error: test failed
```

Notice that using take_until_either with the same parameter makes the test pass.

**Comment by Geal** [comment]
ok, I have found why. `take_until` can return an empty slice, while `take_until_either` does not. I am not sure yet what I'll do about this issue. My first feeling would be to accept empty strings, but I am not convinced that's the right way.

**Comment by seppo0010** [comment]
I feel like empty strings should be acceptable, although that would break my use 😢, although I don't fully understand why.

**Comment by tmccombs** [comment]
I think they should accept empty strings. But maybe there should also be complementing macros that don't accept empty strings (like many1 compared to many0).

**Comment by Geal** [comment]
yes, I think that's the way to go. Fixing `take_until_either!` and `take_until_either_and_consume!` so they can return empty slices, and make `take_until_*1` versions that return errors in that case.

**Comment by Geal** [comment]
Hello, this should be fixed by 484f672
